### ISSUE TITLE
Reduce duplication of data in our schema

### ISSUE THREAD

**Issue by j-rewerts** [canada-ca/ore-ero#279]
We've decided against going with Netlify, but one of the best things about it is the relation widget. We currently use this same type of logic with the administrations drop down at the top of our code form. 

There are several areas where it might help to use a relation rather than storing the data inside 1 large file:

[Contacts](https://github.com/canada-ca/ore-ero/blob/master/_data/schemaCode.yaml#L6) - we see contacts pop up in several places. Maybe we could have separate data files per contact?

[Partners](https://github.com/canada-ca/ore-ero/blob/master/_data/schemaCode.yaml#L36) - Can we make our administrations generic enough to support this just being a list of relations?

[Related Code](https://github.com/canada-ca/ore-ero/blob/master/_data/schemaCode.yaml#L53) - Could this be from oss or code?

I'm sure there are others as well.

The core idea of this is [normalizing our data](https://en.wikipedia.org/wiki/Database_normalization).

**Comment by j-rewerts** [comment]
@smellems @fredericlestage-vdm What are your thoughts? I believe this is the biggest missing piece between our current site and how @WJXHenry and I set up Netlify.

**Comment by fredericlestage-vdm** [comment]
I agree. Even if we're not using a database, it still make sense to think of our yaml files the same way. You wouldn't set up a repeating field (like releases in the code schema) in a database, it wouldn't work. Instead you would link rows in different tables to each other using an id of some sort. It might be a good idea to draw a diagram (inspired by UML) to decide what the relations are between all the elements before we go forward with our changes since it will only get more complex with time.

**Comment by j-rewerts** [comment]
What if we merge OSS and code? Are they really that different? Is our criteria that `code` is managed by one of the organizations that are part of ORE and with OSS, we have no real owner that we can contact?

**Comment by fredericlestage-vdm** [comment]
It might be a good idea to see if they are different enough to justify them being separate elements. However, it wouldn't resolve the issue of large files, since it would only be changing files from one folder to another.

**Comment by j-rewerts** [comment]
Yeah I'm just drawing up a UML diagram, but with just types and multiplicity. I'll make one of OSS with the administrations and contacts separate.

**Comment by j-rewerts** [comment]
Okay so here's a diagram that is similar to what we have currently:

![current](https://user-images.githubusercontent.com/8295080/59523059-84bca800-8e8d-11e9-8d46-42f1ace74c4a.jpg)

I've added contact and made some references more formal. Also, all situations where there is an organization or team, I've chosen to make those `Administrations`. If some sort of hierarchy is needed, we use the `Administrations.parentOrg` field to represent that relationship. Note that the composition relationship is meant to show that there isn't a `uses` table for this. It's contained in `OSS` files. As mentioned previously, this isn't properly normalized, but we could normalize if we wanted.

The main problem with this approach is it isn't clear how to relate code. For example, `Uses (old).relatedCode` I've set to point to `Code`, but maybe it should point to `OSS`. This is the current problem we have with `OSS` vs `Code`. I propose the following:

![Alt (relation)](https://user-images.githubusercontent.com/8295080/59523246-06143a80-8e8e-11e9-995a-dadb0db4fcd9.jpg)

In this diagram, we have a `Uses` table. It relates administrations and the `Code` they use. I think this hits on the main use case of `OSS` before, which is showing who uses what in Government, and giving us someone to talk to if we need it. In this situation, `Code` includes things like WET, YouCanBenefit, Docker, MySQL and any other open source projects out there. There's no line drawn between whether it was created by the government or not.

Let me know what your thoughts are!

**Comment by j-rewerts** [comment]
@fredericlestage-vdm I forgot to respond to your second point. Each of the classes in the UML diagram have their own folder. Each object/row is its own file. This coupled with the fact that we're using relations, rather than embedding data in each file should make things fairly efficient.

**Comment by j-rewerts** [comment]
Also @smellems let me know if I'm off on the primary use of `OSS`. Maybe I've misunderstood it?

**Comment by fredericlestage-vdm** [comment]
Yesterday, I played around in a custom repo (https://github.com/fredericlestage-vdm/ore-ero-db) where I tried splitting our data in multiple files. Here's what I gathered:

- Splitting data from .yml files is quite simple. Even the liquid markup isn't too much complex to do. However, I'm quite certain what I did can be improved upon and instead of using nested for loops, there's probably a way of using filters (I'm specifically thinking about the where filter, which will make it even closer to a sql call)
- We should agree on a folder structure. 
  - Either filing the admins, codes, softwares under their corresponding org level (federal, provincial, municipal, etc.). Would it make sense for a code or a software to be used or created by administrations at different gov. level, if so, this hierarchy doesn't make sense. 
  - Or just having them in their project type root folder (under admin, code or software).
  - Another way would be to use their broad categories or taxonomies (https://docs.google.com/spreadsheets/d/19ZvRwmYufeH6bG41CHnJhUowwx9USGaYi8KDSHysCYM/edit#gid=2104562770). There was supposed to be a discussion about adding these to the projects, but we could also use them as a filing system.
- I'm not so sure about this one, since I lack a bit of knowledge about the core of javascript and file loading, but one of the problem I think we might run into is that for fetching data when will auto-fill the forms, we're still using a single json file that basically just parse the yml folders. I don't think that loading a single large file is necessarily the optimal choice versus multiple small ones, especially if we don't need most of the data we're loading. I don't suppose reading is much of a problem since it's json and it use a key/value pair, but I do think that loading only the data we require would make for better performances. So we might want to split that too.
  - However, the problem I see with that is that we would have to create a new file for each yml data file and use the jsonify filter on each one of them. Unless we find a way to automatically create them when we create a new yml data file, this will be a long process (even just for creating the ones we have for now will take some time)
